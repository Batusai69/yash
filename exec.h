/* Yash: yet another shell */
/* exec.h: command execution */
/* (C) 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef EXEC_H
#define EXEC_H

#include <stdbool.h>
#include <sys/types.h>


#define EXIT_FAILURE1 1
#define EXIT_ERROR    2
#define EXIT_NOEXEC   126
#define EXIT_NOTFOUND 127
#define EXIT_SYNERROR (256 + EXIT_ERROR)
#define EXIT_EXPERROR EXIT_ERROR
#define EXIT_ASSGNERR EXIT_ERROR
#define EXIT_REDIRERR EXIT_ERROR

extern int laststatus;
extern pid_t lastasyncpid;

struct execinfo;
extern void reset_execinfo(void);
extern struct execinfo *save_execinfo(void)
    __attribute__((malloc,warn_unused_result));
extern void load_execinfo(struct execinfo *save)
    __attribute__((nonnull));
extern bool return_pending(void)
    __attribute__((pure));

struct and_or_T;
struct wordunit_T;
extern void exec_and_or_lists(const struct and_or_T *a, bool finally_exit);
extern wchar_t *exec_command_substitution(const wchar_t *code)
    __attribute__((nonnull,malloc,warn_unused_result));
extern int open_heredocument(const struct wordunit_T *content);

extern int return_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern int break_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern int eval_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern int dot_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern int exec_builtin(int argc, void **argv)
    __attribute__((nonnull));
extern const char return_help[], break_help[], continue_help[],
       eval_help[], dot_help[], exec_help[];


#endif /* EXEC_H */


/* vim: set ts=8 sts=4 sw=4 noet: */
