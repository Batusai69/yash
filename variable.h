/* Yash: yet another shell */
/* variable.h: interface to shell variable manager */
/* © 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


extern char **environ;

#define VAR_USER            "USER"
#define VAR_LOGNAME         "LOGNAME"
#define VAR_HOME            "HOME"
#define VAR_LANG            "LANG"
#define VAR_PATH            "PATH"
#define VAR_PWD             "PWD"
#define VAR_OLDPWD          "OLDPWD"
#define VAR_IFS             "IFS"
#define VAR_SHLVL           "SHLVL"
#define VAR_HOSTNAME        "HOSTNAME"
#define VAR_POSIXLY_CORRECT "POSIXLY_CORRECT"
#define VAR_LC_ALL          "LC_ALL"
#define VAR_LC_COLLATE      "LC_COLLATE"
#define VAR_LC_CTYPE        "LC_CTYPE"
#define VAR_LC_MESSAGES     "LC_MESSAGES"
#define VAR_LC_MONETARY     "LC_MONETARY"
#define VAR_LC_NUMERIC      "LC_NUMERIC"
#define VAR_LC_TIME         "LC_TIME"
#define VAR_LINES           "LINES"
#define VAR_COLUMNS         "COLUMNS"

/* 変数を表す構造体 */
struct variable {
	char *value;
	/*union {
		int    as_int;
		double as_double;
	} avalue;*/
	enum {
		VF_EXPORT   = 1 << 0,
		VF_READONLY = 1 << 1,
		VF_NODELETE = 1 << 2,
	} flags;
	const char *(*getter)(struct variable *var);
	bool (*setter)(struct variable *var);
};
/* VF_EXPORT フラグが立っている変数は、export 対象であり、value メンバの内容
 * よりも environ に入っている内容を優先する */
/* getter/setter/deleter は変数のゲッター・セッター・削除関数である。
 * 非 NULL なら、値を取得・設定する際にフィルタとして使用する。
 * ゲッター・セッターの戻り値はそのまま getvar/setvar の戻り値になる。
 * セッターは unsetvar でも value = NULL で呼ばれる。 */

/* 変数代入を元に戻すためのデータ */
struct save_assignment;

char *xgetcwd(void);
void init_var(void);
void set_shlvl(int change);
void set_positionals(char *const *values)
	__attribute__((nonnull));
const char *getvar(const char *name)
	__attribute__((nonnull));
bool setvar(const char *name, const char *value, bool export)
	__attribute__((nonnull));
struct hasht *get_variable_table(void);
struct plist *getarray(const char *name);
bool unsetvar(const char *name)
	__attribute__((nonnull));
bool export(const char *name)
	__attribute__((nonnull));
void unexport(const char *name)
	__attribute__((nonnull));
bool is_exported(const char *name)
	__attribute__((nonnull));
bool assign_variables(char **assigns, bool temp, bool export)
	__attribute__((nonnull));
void unset_temporary(const char *name);
bool is_special_parameter_char(char c);
bool is_name_char(char c);
bool is_name(const char *c)
	__attribute__((nonnull));
