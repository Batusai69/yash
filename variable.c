/* Yash: yet another shell */
/* variable.c: shell variable manager */
/* © 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#include <error.h>
#include <errno.h>
#include <limits.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include "yash.h"
#include "util.h"
#include "expand.h"
#include "exec.h"
#include "variable.h"
#include <assert.h>


/* 変数を表す構造体 */
struct variable {
	char *value;
	/*union {
		int    as_int;
		double as_double;
	} avalue;*/
	enum {
		VF_EXPORT   = 1 << 0,
		VF_READONLY = 1 << 1,
	} flags;
	const char *(*getter)(struct variable *var);
	bool (*setter)(struct variable *var, const char *value, bool export);
};
/* VF_EXPORT フラグが立っている変数は、export 対象であり、value メンバの内容
 * よりも environ に入っている内容を優先する */
/* getter/setter は変数のゲッター・セッターである。
 * 非 NULL なら、値を取得・設定する際にフィルタとして使用する。
 * ゲッター・セッターの戻り値はそのまま getvar/setvar の戻り値になる。
 * ゲッターが非 NULL のとき、value は NULL である。
 * セッターは、variable 構造体の value メンバを書き換える等、変数を設定するのに
 * 必要な動作を全て行う。セッターが NULL ならデフォルトの動作をする。 */

/* 変数環境を表す構造体 */
struct environment {
	struct environment *parent;  /* 親環境 */
	struct hasht variables;      /* 普通の変数のハッシュテーブル */
	struct plist positionals;    /* 位置パラメータのリスト */
};
/* variables の要素は struct variable へのポインタ、positionals の要素は
 * char へのポインタである。 */
/* $0 は位置パラメータではない。故に positionals.contents[0] は NULL である。 */


char *xgetcwd(void);
void init_var(void);
void set_shlvl(int change);
void set_positionals(char *const *values);
static struct variable *get_variable(const char *name, bool temp);
const char *getvar(const char *name);
bool setvar(const char *name, const char *value, bool export);
struct plist *getarray(const char *name);
bool unsetvar(const char *name);
bool export(const char *name);
void unexport(const char *name);
bool is_exported(const char *name);
bool assign_variables(char **assigns, bool temp, bool export);
void unset_temporary(const char *name);
bool is_special_parameter_char(char c);
bool is_name_char(char c);
bool is_name(const char *c);
static const char *count_getter(struct variable *var);
static const char *laststatus_getter(struct variable *var);
static const char *pid_getter(struct variable *var);
static const char *last_bg_pid_getter(struct variable *var);
static const char *zero_getter(struct variable *var);


/* 変数を保存する環境 */
static struct environment *current_env;
/* 一時的変数を保存するハッシュテーブル */
static struct hasht temp_variables;


/* getcwd(3) の結果を新しく malloc した文字列で返す。
 * エラー時は NULL を返し、errno を設定する。 */
char *xgetcwd(void)
{
	size_t pwdlen = 40;
	char *pwd = xmalloc(pwdlen);
	while (getcwd(pwd, pwdlen) == NULL) {
		if (errno == ERANGE) {
			pwdlen *= 2;
			pwd = xrealloc(pwd, pwdlen);
		} else {
			free(pwd);
			pwd = NULL;
			break;
		}
	}
	return pwd;
}


/* 環境変数などを初期化する。 */
void init_var(void)
{
	struct variable *var;

	current_env = xmalloc(sizeof *current_env);
	current_env->parent = NULL;
	ht_init(&current_env->variables);
	pl_init(&current_env->positionals);
	pl_append(&current_env->positionals, NULL);

	ht_init(&temp_variables);

	/* まず current_env->variables に全ての環境変数を設定する。 */
	for (char **e = environ; *e; e++) {
		size_t namelen = strcspn(*e, "=");
		char name[namelen + 1];
		strncpy(name, *e, namelen);
		name[namelen] = '\0';
		if ((*e)[namelen] != '=')
			continue;

		struct variable *var = xmalloc(sizeof *var);
		*var = (struct variable) {
			.value = xstrdup(*e + namelen + 1),
			.flags = VF_EXPORT,
		};
		ht_set(&current_env->variables, name, var);
	}

	/* 特殊パラメータを設定する。 */
	var = xmalloc(sizeof *var);
	*var = (struct variable) {
		.value = NULL,
		.getter = count_getter,
	};
	ht_set(&current_env->variables, "#", var);
	var = xmalloc(sizeof *var);
	*var = (struct variable) {
		.value = NULL,
		.getter = laststatus_getter,
	};
	ht_set(&current_env->variables, "?", var);
	/*var = xmalloc(sizeof *var);
	*var = (struct variable) {
		.value = NULL,
		.getter = TODO $- parameter,
	};
	ht_set(&current_env->variables, "-", var);*/
	var = xmalloc(sizeof *var);
	*var = (struct variable) {
		.value = NULL,
		.getter = pid_getter,
	};
	ht_set(&current_env->variables, "$", var);
	var = xmalloc(sizeof *var);
	*var = (struct variable) {
		.value = NULL,
		.getter = last_bg_pid_getter,
	};
	ht_set(&current_env->variables, "!", var);
	var = xmalloc(sizeof *var);
	*var = (struct variable) {
		.value = NULL,
		.getter = zero_getter,
	};
	ht_set(&current_env->variables, "0", var);

	/* デフォルトの export を設定する。 */
	export(VAR_HOME);
	export(VAR_LANG);
	export(VAR_LC_ALL);
	export(VAR_LC_COLLATE);
	export(VAR_LC_CTYPE);
	export(VAR_LC_MESSAGES);
	export(VAR_LC_MONETARY);
	export(VAR_LC_NUMERIC);
	export(VAR_LC_TIME);
	export(VAR_PATH);
	export(VAR_PWD);
	export(VAR_OLDPWD);
	export(VAR_SHLVL);
	export(VAR_LINES);
	export(VAR_COLUMNS);

	/* PWD 環境変数を設定する */
	char *pwd = xgetcwd();
	if (pwd) {
		setvar(VAR_PWD, pwd, true);
		free(pwd);
	}
}

/* 現在の環境の位置パラメータを設定する。既存の位置パラメータは削除する。
 * values[0] が $1 に、values[1] が $2 に、という風になる。values[x] が NULL
 * になったら終わり。 */
void set_positionals(char *const *values)
{
	struct plist *pos = &current_env->positionals;

	/* まず古いのを消す。 */
	for (size_t i = 1; i < pos->length; i++)
		free(pos->contents[i]);
	pl_remove(pos, 1, SIZE_MAX);

	/* 新しいのを追加する。 */
	assert(pos->length == 1);
	while (*values) {
		pl_append(pos, xstrdup(*values));
		values++;
	}
}

/* 環境変数 SHLVL に change を加える */
void set_shlvl(int change)
{
	const char *shlvl = getvar(VAR_SHLVL);
	int level = shlvl ? atoi(shlvl) : 0;
	char newshlvl[16];

	level += change;
	if (level < 0)
		level = 0;
	if (snprintf(newshlvl, sizeof newshlvl, "%d", level) >= 0) {
		if (!setvar(VAR_SHLVL, newshlvl, false))
			error(0, 0, "failed to set env SHLVL");
	}
}

/* 現在の変数環境や親変数環境・一時的変数から指定した変数を捜し出す。
 * 位置パラメータ、$*, $@ は取得できない。
 * temp: 一時的変数からも探すかどうか。 */
static struct variable *get_variable(const char *name, bool temp)
{
	if (temp) {
		struct variable *var = ht_get(&temp_variables, name);
		if (var)
			return var;
	}

	struct environment *env = current_env;
	while (env) {
		struct variable *var = ht_get(&env->variables, name);
		if (var)
			return var;
		env = env->parent;
	}
	return NULL;
}

/* 指定した名前のシェル変数を取得する。
 * 特殊パラメータ $* および $@ は得られない。
 * 変数が存在しないときは NULL を返す。 */
const char *getvar(const char *name)
{
	struct variable *var = get_variable(name, true);  /* 普通の変数を探す */
	if (var) {
		if (var->getter)
			return var->getter(var);
		if (var->flags & VF_EXPORT)
			return getenv(name);
		return var->value;
	}
	if (xisdigit(name[0])) {  /* 位置パラメータを探す */
		char *end;
		errno = 0;
		size_t posindex = strtoul(name, &end, 10);
		if (*end || errno)
			return NULL;
		if (0 < posindex && current_env->positionals.length)
			return current_env->positionals.contents[posindex];
	}
	return NULL;
}

/* シェル変数の値を設定する。変数が存在しない場合、基底変数環境に追加する。
 * この関数では位置・特殊パラメータは設定できない (してはいけない)。
 * 同名の一時的変数があれば、一時的変数を削除してから普通に変数を設定する。
 * name:   正しい変数名。
 * export: true ならその変数を export 対象にする。false ならそのまま。
 * 戻り値: 成功なら true、エラーならメッセージを出して false。 */
// TODO 配列も代入可能に
bool setvar(const char *name, const char *value, bool export)
{
	assert(!xisdigit(name[0]));
	assert(!is_special_parameter_char(name[0]));

	unset_temporary(name);

	struct variable *var = get_variable(name, false);
	if (!var) {
		struct environment *env = current_env;
		while (env->parent)
			env = env->parent;
		var = xmalloc(sizeof *var);
		*var = (struct variable) {
			.value = NULL,
			.flags = 0,
			.getter = NULL,
			.setter = NULL,
		};
		ht_set(&env->variables, name, var);
	}
	if (var->flags & VF_READONLY) {
		error(0, 0, "%s: readonly variable cannot be assigned to", name);
		return false;
	}

	bool ok;
	if (var->setter) {
		ok = var->setter(var, value, export);
	} else {
		ok = true;
		if (export)
			var->flags |= VF_EXPORT;
		if (var->flags & VF_EXPORT) {
			if (setenv(name, value, true) < 0) {
				error(0, errno, "export %s=%s", name, value);
				ok = false;
			}
		}
		if (var->value != value) {
			free(var->value);
			var->value = xstrdup(value);
		}
	}
	assert(var->value != NULL);
	return ok;
}

/* 指定した名前の配列変数の内容を取得する。
 * name が NULL なら位置パラメータリストを取得する。
 * 配列変数が見付からなければ NULL を返す。 */
/* 位置パラメータのインデックスは 0 ではなく 1 から始まることに注意 */
struct plist *getarray(const char *name)
{
	if (!name)
		return &current_env->positionals;
	return NULL; //XXX 配列は未実装
}

/* シェル変数を削除する
 * 戻り値: エラーがなければ true、変数を削除できなければ false。
 *         変数が存在しなかったら true。 */
bool unsetvar(const char *name)
{
	unset_temporary(name);

	struct environment *env = current_env;
	while (env) {
		struct variable *var = ht_remove(&env->variables, name);
		if (var) {
			bool oldvarexport = var->flags & VF_EXPORT;
			if (var->flags & VF_READONLY) {
				ht_set(&env->variables, name, var);
				error(0, 0, "cannot unset readonly variable %s", name);
				return false;
			}
			free(var->value);
			free(var);

			/* 親環境に同名の変数があれば、environ の内容をそれに合わせる */
			var = get_variable(name, false);
			if (!var || !(var->flags & VF_EXPORT))
				unsetenv(name);
			else if (!getenv(name) && oldvarexport)
				setenv(name, var->value, true);
			return true;
		}
		env = env->parent;
	}
	return true;
}

/* 変数を export 対象にする。
 * 戻り値: 成功したかどうか */
bool export(const char *name)
{
	unset_temporary(name);

	struct variable *var = get_variable(name, false);
	if (var) {
		if (setenv(name, var->value, false) < 0) {
			error(0, errno, "export %s=%s", name, var->value);
			return false;
		}
	} else {
		struct environment *env = current_env;
		while (env->parent)
			env = env->parent;
		var = xmalloc(sizeof *var);
		*var = (struct variable) {
			.value = NULL,
			.flags = VF_EXPORT,
		};
		ht_set(&env->variables, name, var);
	}
	return true;
}

/* 変数を非 export 対象にする。 */
void unexport(const char *name)
{
	unset_temporary(name);

	struct variable *var = get_variable(name, false);
	if (var)
		var->flags &= ~VF_EXPORT;
	unsetenv(name);
}

/* 指定した名前のシェル変数が存在しかつ export 対象かどうかを返す。
 * 一時的変数は考慮しない。 */
bool is_exported(const char *name)
{
	struct variable *var = get_variable(name, false);
	return var && (var->flags & VF_EXPORT);
}

/* 変数代入を行う。
 * assigns: 代入する "変数=値" の配列へのポインタ。
 * temp:    true なら一時的変数として代入する。false なら普通に代入する。
 * export:  代入する変数を export 対象にするかどうか。
 * 戻り値:  成功すれば true、エラーなら false。
 * temp と export を両方 true にしてはいけない。
 * エラーになっても途中結果を元に戻す為のデータが *save に入る。 */
bool assign_variables(char **assigns, bool temp, bool export)
{
	assert(!temp || !export);

	char *assign;
	while ((assign = *assigns)) {
		//XXX 配列の代入は未対応
		size_t namelen = strcspn(assign, "=");
		char name[namelen + 1];
		strncpy(name, assign, namelen);
		name[namelen] = '\0';
		assert(namelen > 0 && assign[namelen] == '=');

		char *value = expand_word(assign + namelen + 1, true);
		if (!value) {
			return false;
		}
		if (temp) {
			struct variable *var = xmalloc(sizeof *var);
			*var = (struct variable) { .value = value, .flags = 0, };
			var = ht_set(&temp_variables, name, var);
			if (var) {
				free(var->value);
				free(var);
			}
		} else {
			if (!setvar(name, value, export)) {
				free(value);
				return false;
			}
		}
		assigns++;
	}
	return true;
}

/* unset_temporary で使う内部関数 */
static int free_tempvar(const char *name __attribute__((unused)), void *value)
{
	struct variable *var = value;
	assert(var != NULL);
	free(var->value);
	free(var);
	return 0;
}

/* 一時的変数を削除する。
 * name: 削除する一時的変数の名前。NULL なら全部の一時的変数を削除する。 */
void unset_temporary(const char *name)
{
	if (name) {
		struct variable *var = ht_remove(&temp_variables, name);
		if (var) {
			free(var->value);
			free(var);
		}
	} else {
		ht_each(&temp_variables, free_tempvar);
		ht_clear(&temp_variables);
	}
}

/* 引数 c が特殊パラメータの名前であるかどうか判定する。
 * 例外的に、'0' はこの関数では false を返す。 */
bool is_special_parameter_char(char c)
{
	return strchr("@*#?-$!_", c) != NULL;
}

/* 文字が変数名に使えるかどうか判定する。 */
bool is_name_char(char c)
{
	switch (c) {
		case 'a':  case 'b':  case 'c':  case 'd':  case 'e':  case 'f':
		case 'g':  case 'h':  case 'i':  case 'j':  case 'k':  case 'l':
		case 'm':  case 'n':  case 'o':  case 'p':  case 'q':  case 'r':
		case 's':  case 't':  case 'u':  case 'v':  case 'w':  case 'x':
		case 'y':  case 'z':  case '_':
		case 'A':  case 'B':  case 'C':  case 'D':  case 'E':  case 'F':
		case 'G':  case 'H':  case 'I':  case 'J':  case 'K':  case 'L':
		case 'M':  case 'N':  case 'O':  case 'P':  case 'Q':  case 'R':
		case 'S':  case 'T':  case 'U':  case 'V':  case 'W':  case 'X':
		case 'Y':  case 'Z':  case '0':  case '1':  case '2':  case '3':
		case '4':  case '5':  case '6':  case '7':  case '8':  case '9':
			return true;
		default:
			return false;
	}
}

/* 文字列が変数名として正しいかどうか判定する。 */
bool is_name(const char *s)
{
	if ('0' <= *s && *s <= '9')
		return false;
	while (is_name_char(*s)) s++;
	return !*s;
}

/* 特殊パラメータ $# のゲッター。位置パラメータの数を返す。 */
static const char *count_getter(
		struct variable *var __attribute__((unused)))
{
	static char result[INT_STRLEN_BOUND(size_t) + 1];
	if (snprintf(result, sizeof result, "%zd",
				current_env->positionals.length - 1) >= 0)
		return result;
	return NULL;
}

/* 特殊パラメータ $? のゲッター。laststatus の値を返す。 */
static const char *laststatus_getter(
		struct variable *var __attribute__((unused)))
{
	static char result[INT_STRLEN_BOUND(int) + 1];
	if (snprintf(result, sizeof result, "%d", laststatus) >= 0)
		return result;
	return NULL;
}

/* 特殊パラメータ $$ のゲッター。シェルのプロセス ID の値を返す。 */
static const char *pid_getter(
		struct variable *var __attribute__((unused)))
{
	static char result[INT_STRLEN_BOUND(pid_t) + 1];
	if (snprintf(result, sizeof result, "%jd", (intmax_t) shell_pid) >= 0)
		return result;
	return NULL;
}

/* 特殊パラメータ $! のゲッター。最後に起動したバックグラウンドジョブの PID。 */
static const char *last_bg_pid_getter(
		struct variable *var __attribute__((unused)))
{
	static char result[INT_STRLEN_BOUND(pid_t) + 1];
	if (snprintf(result, sizeof result, "%.0jd", (intmax_t) last_bg_pid) >= 0)
		return result;
	return NULL;
}

/* 特殊パラメータ $0 のゲッター。実行中のシェル(スクリプト)名を返す。 */
static const char *zero_getter(
		struct variable *var __attribute__((unused)))
{
	return command_name;
}
