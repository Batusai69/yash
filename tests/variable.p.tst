echo $LINENO
cat <<EOF
some
here
document
EOF
cat <<\END
another
here-document
END
echo $LINENO
echo line \
continuation
echo $LINENO

printlineno () {
	echo function
	echo $LINENO
}
printlineno
echo $LINENO

echo ===== 1 =====

var=var
echo $var
var='v'"a"r
echo $var
var=xxx echo $var

foo='123  456  789'
bar=-${foo}-
echo $bar
echo "$bar"

echo ===== 2 =====

func () {
	var=xyz
	echo $var
}
func
echo $var

echo ===== set export =====

export save="$(env -i LC_CTYPE="$LC_CTYPE" foo=123 $INVOKE $TESTEE -c 'bar=456; set')"
$INVOKE $TESTEE -c 'eval $save; echo $foo $bar'

unset save
echo - $save -
echo save ${save-unset}

export var
setvar="$(set | grep '^var=')"
var=123
echo $var
eval "$setvar"
echo $var
exportvar="$(export | grep -E '^export[[:blank:]]+var=')"
unset var
echo var ${var-unset}
eval "$exportvar"
$INVOKE $TESTEE -c 'echo $var'

unset null
export null
exportnull="$(export | grep -E '^export[[:blank:]]+null$')"
unset null
eval "$exportnull"
null=null
$INVOKE $TESTEE -c 'echo $null'

echo ===== readonly =====

readonly var ro=readonly! empty
echo $var $ro -$empty-
(var=no; echo $var) 2>/dev/null
(empty=no; echo -$empty-) 2>/dev/null
(unset var; echo $var) 2>/dev/null
(var=no sleep 0 && echo not reached - sleep) 2>/dev/null
(var=no cd && echo not reached - cd) 2>/dev/null
export readonlyvar="$(readonly | grep -E '^readonly[[:blank:]]+var=')"
$INVOKE $TESTEE -s <<\END
eval $readonlyvar
(var=no; echo $var) 2>/dev/null
END
export readonlyempty="$(readonly | grep -E '^readonly[[:blank:]]+empty$')"
$INVOKE $TESTEE -s <<\END
eval $readonlyempty
(empty=no; echo -$empty-) 2>/dev/null
END

echo ===== function unset =====

echo () { :; }
echo ng
unset -f echo
echo ok

echo ===== set shift =====

set 1 2 3 a b c
shift
echo "$@"
shift 2
echo "$@"
shift 4 2>/dev/null || echo shift 4

loop ()
while [ $# -ne 0 ]; do
	echo "$@"
	shift
done
loop x y z
echo "$@"
shift 3
echo $# "$@"

echo ===== getopts =====

echo $OPTIND

set -- -a -b "foo  bar" -c - -- -d baz
while getopts ab:c opt
do echo $opt "${OPTARG-unset}"
done
echo $opt $OPTIND

echo =====
OPTIND=1
while getopts a:b opt -a -- -abc -b -- -a foo
do echo $opt "${OPTARG-unset}"
done
echo $opt $OPTIND

echo =====
OPTIND=1
while getopts :abcde opt -abc -de fgh
do echo $opt "${OPTARG-unset}"
done
echo $opt $OPTIND

echo =====
set --
OPTIND=1
! getopts "" opt
echo $? $opt $OPTIND "${OPTARG-unset}"

echo =====
set -- -a b c
OPTIND=1
getopts "" opt 2>/dev/null
echo $? $opt "${OPTARG-unset}"

set -- -a b c
OPTIND=1
getopts : opt
echo $? $opt "${OPTARG-unset}"

set -- -a
OPTIND=1
getopts a: opt 2>/dev/null
echo $? $opt "${OPTARG-unset}"

set -- -a
OPTIND=1
getopts :a: opt
echo $? $opt "${OPTARG-unset}"

echo ===== read =====

unset IFS a b c d e

read a b c <<END
1 2 3
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"

read a b c <<END
1 2 3 4  5
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"

read a b c <<END
1 2 3
4 5
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"

read a b c <<END
1 2 3\
4  5
6 7
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"

read a b c d e <<END
1 2 3
4 5
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"
echo "${d-error}"
echo "${e-error}"

echo =====

IFS=- read a b c <<END
1 2\-3-4 5\-6-7\-8-9
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"

IFS=- read -r a b c <<\END
1 2\-3-4 5\-6-7\-8-9\
0
END
echo "${a-error}"
echo "${b-error}"
echo "${c-error}"

while read a b
do
	echo $b
done <<END
1 2 3 4
5 6 7 8
9 0 1 2
END

echo ${IFS-unset}
