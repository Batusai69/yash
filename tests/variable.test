tmp=/tmp/yash-variable-test-$$

e\
c\
h\
o $LINENO

RANDOM=123
(echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM
RANDOM=456
echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM) >$tmp-1
(echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM
RANDOM=456
echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM) >$tmp-2
diff $tmp-1 $tmp-2 && echo ok

echo ===== 1 =====

func1 () { echo func1; }
func2 () { echo func2; }

func1; func2;

func1 () {
	func1 () { echo re-defined func1; }
	func2 () { echo re-defined func2; }
	echo re-defined.
}
func2; func1; func2; func1;


# TODO should be removed in EXIT trap
rm -f $tmp-1 $tmp-2
