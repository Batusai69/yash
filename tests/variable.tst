RANDOM=123
(echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM
RANDOM=456
echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM) >"${TESTTMP}/variable-a"
(echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM
RANDOM=456
echo $RANDOM $RANDOM $RANDOM $RANDOM $RANDOM) >"${TESTTMP}/variable-b"
diff "${TESTTMP}/variable-a" "${TESTTMP}/variable-b" && echo ok

oldrand=$RANDOM rand=$RANDOM
while [ "$oldrand" = "$rand" ]; do
	oldrand=$rand rand=$RANDOM
done

echo ===== 1 =====

func1 () { echo func1; }
func2 () { echo func2; }

func1; func2;

func1 () {
	func1 () { echo re-defined func1; }
	func2 () { echo re-defined func2; }
	echo re-defined.
}
func2; func1; func2; func1;

echo ===== 2 =====

echol () {
	typeset i  # local
	for i
	do printf "%s\n" "$i"
	done
}

unset i
ary=123
echol "$ary"
ary=(1 22 '3  3' 4\ \ \ 4 5)
echol "$ary"
ary=456
echol "$ary"
ary=(9\
9
8'
'8"
"8
7 #comment
${ary/5/ })
echol "$ary"
ary=(test of array)
echo i "${i-unset}"

ary0=()
unset ary0
echo ary0 "${ary0-unset}"

echo ===== typeset export =====

func () {
	typeset foo=abc bar
	bar=def
	echo $foo$bar
	export foo bar=xyz
	$INVOKE $TESTEE -c 'echo $foo$bar'
	typeset -g baz qux=global
}
foo=foo bar=bar func
$INVOKE $TESTEE -c 'echo $foo $bar'
echo ${baz-unset} $qux

func () {
	export -g baz qux
	export -p baz qux
}
func
$INVOKE $TESTEE -c 'echo ${baz-unset} $qux'

typeset -px baz qux

typeset -X qux
$INVOKE $TESTEE -c 'echo ${qux-unset}'
echo ${qux-unset}

export ary
export -p ary
env | grep "^ary="

echo ===== unset readonly =====

unset foo bar baz qux
$INVOKE $TESTEE -c 'echo ${foo-unset} ${bar-unset} ${baz-unset} ${qux-unset}'

readonly ro=readonly
echo $ro
func () {
	typeset ro=local
	echo func $ro
	unset ro
	echo func $ro
	typeset ro
	ro=localagain
	echo func $ro
	readonly -x ro2=export
	$INVOKE $TESTEE -c 'echo $ro2'
	(ro2=xxx; unset ro2; $INVOKE $TESTEE -c 'echo $ro2') 2>/dev/null
	readonly -p
}
func
unset ro3
ro3=dummy typeset -r ro3=ro3

typeset -pr ro ro3

for num in 1 2 3 4 5; do
	echo $num
	if [ $num = 3 ]; then
		readonly num
	fi
done 2>/dev/null

readonly ary
ary=(cannot assign) 2>/dev/null
readonly -p ary
typeset -p ary
echo $ary

echo ===== function typeset =====

func () {
	echo ok
}
typeset -fp func
typeset -fp

readonly -f func
{
	func () { echo invalid re-definition; }
} 2>/dev/null
func

echo ===== read =====

unset a
printf "%s" '12\' | (
read a
echo $? "$a"
)
printf "%s" '12\' | (
read -r a
echo $? "$a"
)


rm -f "${TESTTMP}/variable-a" "${TESTTMP}/variable-b"
