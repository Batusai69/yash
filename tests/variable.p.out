1
some
here
document
another
here-document
11
line continuation
14
function
18
21
===== 1 =====
var
var
var
-123 456 789-
-123  456  789-
===== 2 =====
xyz
xyz
===== 3 =====
1 $0=variable.p.tst
2 $0=sh
3 $0=sh
4 $0=command name
===== set export =====
123 456
- -
save unset
123
xyz
var unset
xyz
null
===== readonly =====
xyz readonly! --
xyz
--
xyz
xyz
--
===== function unset =====
ok
===== set shift =====
2 3 a b c
a b c
shift 4
x y z
y z
z
a b c
0
===== getopts =====
1
a unset
b foo  bar
c unset
? 5
===== 1
a --
a bc
b unset
? 6
===== 2
a unset
b unset
c unset
d unset
e unset
? 3
a unset
b unset
c unset
d unset
e unset
? 3
===== 3
0 ? 1 unset
===== 4
0 ? unset
0 ? a
0 ? unset
0 : a
===== read =====
1
2
3
1
2
3 4  5
1
2
3
1
2
34  5
1
2
3


=====
1 2-3
4 5-6
7-8-9
1 2\
3
4 5\-6-7\-8-9\
2 3 4
6 7 8
0 1 2
unset
